frontend/heir/mlir/types.py

```diff
@@ -34,115 +36,128 @@ def __mul__(self, other) -> Self:
 class Secret(Generic[T], MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
+  def numba_type() -> NumbaType:
     raise NotImplementedError("No numba type exists for a generic Secret")
 
 
 class Tensor(Generic[*Ts], MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
+  def numba_type() -> NumbaType:
     raise NotImplementedError("No numba type exists for a generic Tensor")
 
 
 class F32(MLIRTypeAnnotation):
   # TODO (#1162): For CKKS/Float: allow specifying actual intended precision/scale and warn/error if not achievable  @staticmethod
   @staticmethod
-  def numba_str():
-    return "float32"
+  def numba_type() -> NumbaType:
+    return float32
 
 
 class F64(MLIRTypeAnnotation):
   # TODO (#1162): For CKKS/Float: allow specifying actual intended precision/scale and warn/error if not achievable  @staticmethod
   @staticmethod
-  def numba_str():
-    return "float64"
+  def numba_type() -> NumbaType:
+    return float64
 
 
 class I1(MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
-    return "bool"
-
-
-class I4(MLIRTypeAnnotation):
-
-  @staticmethod
-  def numba_str():
-    return "int4"
+  def numba_type() -> NumbaType:
+    return boolean
 
 
 class I8(MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
-    return "int8"
+  def numba_type() -> NumbaType:
+    return int8
 
 
 class I16(MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
-    return "int16"
+  def numba_type() -> NumbaType:
+    return int16
 
 
 class I32(MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
-    return "int32"
+  def numba_type() -> NumbaType:
+    return int32
 
 
 class I64(MLIRTypeAnnotation):
 
   @staticmethod
-  def numba_str():
-    return "int64"
+  def numba_type() -> NumbaType:
+    return int64
 
 
 # Helper functions
 
 
-def to_numba_str(type) -> str:
+def to_numba_type(type: type) -> NumbaType:
   if get_origin(type) == Secret:
     raise TypeError(
         "Secret type should not appear inside another type annotation."
     )
 
   if get_origin(type) == Tensor:
     args = get_args(type)
+    if len(args) < 2:
+      raise TypeError(
+          "Tensor should contain a type and at least one dimension, but found"
+          f" {type}"
+      )
     inner_type = args[-1]
     if get_origin(inner_type) == Tensor:
       raise TypeError("Nested Tensors are not yet supported.")
+    # This is slightly cursed, as numba constructs array types via slice syntax
     # Cf. https://numba.pydata.org/numba-doc/dev/reference/types.html#arrays
-    return f"{to_numba_str(inner_type)}[{','.join([':'] * (len(args) - 1))}]"
+    ty = to_numba_type(inner_type)[(slice(None),) * (len(args) - 1)]
+    # We augment the type object with `shape` for the actual sizes
+    ty.shape = args[:-1]  # type: ignore
```

Comment:
Possible to have the shape here be its own tuple/list?

`Tensor[(2,3,4), I16]` instead of mixing the member type and the shape? I think that would be slightly more Pythonic.